HW2/P2/parallel_vector.pyx

```diff
@@ -94,22 +129,62 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
                                np.int32_t[:] dest,
                                int repeat,
                                int N):
-   cdef:
-       int idx, r
-       int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
-       omp_lock_t *locks = get_N_locks(num_locks)
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for every N adjacent elements of counts
-   # to parallelize data movement.  Be sure to avoid deadlock, as well as
-   # double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(num_locks, locks)
+    cdef:
+        int idx, r,src_lk,dest_lk
+        int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
+        omp_lock_t *locks = get_N_locks(num_locks)
+ 
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for every N adjacent elements of counts
+    # to parallelize data movement.  Be sure to avoid deadlock, as well as
+    # double-locking.
+    ##########
+    for r in range(repeat):
+        for idx in prange(src.shape[0], nogil=True, num_threads=4):
+            src_lk=src[idx]/num_locks
+            dest_lk=dest[idx]/num_locks
+
+            if src_lk>=num_locks:
+                src_lk=num_locks-1
+            if dest_lk>=num_locks:
+                dest_lk=num_locks-1
+
+            if src_lk<dest_lk:
+                #THIS PART MAKES SURE THAT IT IS IN ORDER AND AVOIDS
+                #DEADLOCKING
+                acquire(&(locks[src_lk]))
+                acquire(&(locks[dest_lk]))
+
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
+
+                release(&(locks[src_lk]))
```

Comment:
Curious to know whether you think there's a difference with releasing the locks in the same order, or release the locks in the reverse order in which you engaged them.